src/messages/MOSDOp.h

```diff
@@ -48,8 +48,10 @@ class MOSDOp : public Message {
   pg_t pgid;
   bufferlist::iterator p;
   // Decoding flags. Decoding is only needed for messages catched by pipe reader.
-  bool partial_decode_needed;
-  bool final_decode_needed;
+  // Transition from true -> false without locks being held
+  // Can never see final_decode_needed == false and partial_decode_needed == true
+  volatile bool partial_decode_needed;
+  volatile bool final_decode_needed;
```

Comment:
I'm just looking up the volatile keyword here, but I'm not sure this works, strictly speaking. http://en.cppreference.com/w/cpp/language/cv and https://en.wikipedia.org/wiki/Volatile_(computer_programming)#In_C_and_C.2B.2B are both pretty clear that ```volatile``` is not "for inter-thread communication"